test/deform/linearized_deform_test.py

```diff
@@ -0,0 +1,302 @@
+    return 4 * x[0] * x[1]
+
+
+def hat_deform_grad0(x, **kwargs):
+    """First component of the deformed gradient of the hat function,
+    which is the same as evaluating the gradient
+    at (x, y) + eps * (2x+y, y+3xy).
+    """
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp = np.where(abs(x0) >= 1e-6, -x0 / abs(x0) * (r - abs(x1)), 0)
+    return tmp * (abs(x0) < r) * (abs(x1) < r)
+
+
+def hat_deform_grad1(x, **kwargs):
+    """Second component of the deformed gradient of the hat function,
+    which is the same as evaluating the gradient
+    at (x, y) + eps * (2x+y, y+3xy).
+    """
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp = np.where(abs(x1) >= 1e-6, -x1 / abs(x1) * (r - abs(x0)), 0)
+    return tmp * (abs(x0) < r) * (abs(x1) < r)
+
+
+def fixed_templ_Deriv_hat(x, **kwargs):
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp1 = np.where(abs(x0) >= 1e-6, -x0 / abs(x0) * (r - abs(x1)), 0)
+    dg0 = tmp1 * (abs(x0) < r) * (abs(x1) < r)
+
+    tmp2 = np.where(abs(x1) >= 1e-6, -x1 / abs(x1) * (r - abs(x0)), 0)
+    dg1 = tmp2 * (abs(x0) < r) * (abs(x1) < r)
+
+    v0 = x[0] - x[1]
+    v1 = 4 * x[0] * x[1]
+    return dg0 * v0 + dg1 * v1
+
+
+def fixed_templ_adj_hat1(x, **kwargs):
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp1 = np.where(abs(x0) >= 1e-6, -x0 / abs(x0) * (r - abs(x1)), 0)
+    dg0 = tmp1 * (abs(x0) < r) * (abs(x1) < r)
+
+    hat = (r - abs(x[0])) * (abs(x[0]) < r) * (r - abs(x[1])) * (abs(x[1]) < r)
+    return dg0 * hat
+
+
+def fixed_templ_adj_hat2(x, **kwargs):
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp = np.where(abs(x1) >= 1e-6, -x1 / abs(x1) * (r - abs(x0)), 0)
+    dg1 = tmp * (abs(x0) < r) * (abs(x1) < r)
+
+    hat = (r - abs(x[0])) * (abs(x[0]) < r) * (r - abs(x[1])) * (abs(x[1]) < r)
+    return dg1 * hat
+
+
+def given_hat(x, **kwargs):
+    r = kwargs.pop('r', 0.5)
+    retval = (r - abs(x[0])) * (abs(x[0]) < r)
+    for xi in x[1:]:
+        retval = retval * (r - abs(xi)) * (abs(xi) < r)
+    return retval
+
+
+def inv_deform_hat(x, **kwargs):
+    """Analytic deformed hat function with support in square with side r.
+    The analytic displacement field (x,y) -> - eps * (x+y, 3xy)."""
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] - eps * (x[0] + x[1])
+    x1 = x[1] - eps * 3 * x[0] * x[1]
+    return (r - abs(x0)) * (abs(x0) < r) * (r - abs(x1)) * (abs(x1) < r)
+
+
+def exp_div_inv_disp(x, **kwargs):
+    eps = 0.25
+    x0 = -(eps + 3 * eps * x[0])
+    return np.exp(x0)
+
+
+# Define the analytic template as the hat function and its gradient
```

Comment:
As said, we do not have global variables in tests since it messes with the testing mechanisms. This needs to be moved to a fixture or moved inside the tests in some other way.